python/lsst/pipe/tasks/photoCal.py

```diff
@@ -657,66 +448,34 @@ def run(self, exposure, sourceCat, expId=0):
             except:
                 self.fig = pyplot.figure()
 
-        if displaySources:
-            frame = 1
-            ds9.mtv(exposure, frame=frame, title="photocal")
-        else:
-            frame = None
-
-        res = self.loadAndMatch(exposure, sourceCat)
-
-        # from res.matches, reserve a fraction of the sources from res.matches and mark the sources reserved
-
-        if self.config.reserveFraction > 0:
-            # Note that the seed can't be set to 0, so guard against an improper expId.
-            random = afwMath.Random(seed=self.config.reserveSeed*(expId if expId else 1))
-            reserveList = []
-            n = len(res.matches)
-            for i in range(int(n*self.config.reserveFraction)):
-                index = random.uniformInt(n)
-                n -= 1
-                candidate = res.matches[index]
-                res.matches.remove(candidate)
-                reserveList.append(candidate)
-
-            if reserveList and self.reservedKey is not None:
-                for candidate in reserveList:
-                    candidate.second.set(self.reservedKey, True)
+        filterName = exposure.getFilter().getName()
 
-        matches = res.matches
-        for m in matches:
-            if self.candidateKey is not None:
-                m.second.set(self.candidateKey, True)
+        # Match sources
+        matchResults = self.match.run(sourceCat, filterName)
+        matches = matchResults.matches
+        reserveResults = self.reserve.run([mm.second for mm in matches], expId=expId)
+        if displaySources:
+            self.displaySources(exposure, matches, reserveResults.reserved)
+        if reserveResults.reserved.sum() > 0:
+            matches = [mm for mm, use in zip(matches, reserveResults.use) if use]
```

Comment:
Why not `matches = matches[reserveResults.use]`?

Reply:
Again, it's a `list`, which requires a listcomp.